### PR TITLE
point-cluster for scattergl

### PR DESCRIPTION
**Pull request by dfcreative**
This PR includes improvement on regl-scatter2d & introduces point-clustering.
Btw selection, panning & updating are significantly faster.
Clustering is somewhat faster.
Introducing async clustering - later.

cc @etpinard

### CODE REVIEW

File: src/traces/scattergl/index.js [hunk 1]
```diff
@@ -882,8 +893,8 @@ function plot(gd, subplot, cdata) {
     if(selectMode) {
         // create select2d
         if(!scene.select2d) {
-            // create scatter instance by cloning scatter2d
-            scene.select2d = createScatter(fullLayout._glcanvas.data()[1].regl, {clone: scene.scatter2d});
+            // create select scatter instance by cloning scatter2d
+            scene.select2d = createScatter(fullLayout._glcanvas.data()[1].regl);
```
Comment (dfcreative) [inline src/traces/scattergl/index.js]: Since we reuse clustering, creating of new scatter is way faster, we don't need to clone data.

**Comment by etpinard** [comment]
Thanks @dfcreative ! We should probably apply those commits on top of https://github.com/plotly/plotly.js/pull/2492 to avoid conflicts.

File: src/traces/scattergl/index.js [hunk 2]
```diff
@@ -271,34 +307,9 @@ function sceneOptions(gd, subplot, trace, positions) {
...
-            errors[p + eOffset[0]] = positions[i * 2 + pOffset] - ax.d2l(vals[i][axLetter + 's']) || 0;
-            errors[p + eOffset[1]] = ax.d2l(vals[i][axLetter + 'h']) - positions[i * 2 + pOffset] || 0;
-            errors[p + eOffset[2]] = 0;
-            errors[p + eOffset[3]] = 0;
-        }
-
-        if(errorOpts.copy_ystyle) {
-            errorOpts = trace.error_y;
-        }
+        if (!trace.unselected) unselectedOptions.opacity = DESELECTDIM;
```
Comment (etpinard) [inline src/traces/scattergl/index.js]: Why do we need all of a sudden?

File: src/traces/scattergl/index.js (hunk 1 above)
Comment (etpinard) [reply src/traces/scattergl/index.js]: sweet!

Comment (etpinard) [reply src/traces/scattergl/index.js]: ... which makes switching from zoom/pan to select/lasso drag modes waaay faster. Thanks!

**Comment by etpinard** [comment]
I'm seeing ~200ms improvements in first-render speed (from 1400-1500 down to ~1200ms) which is pretty amazing :tada: 


That said, on my machine drag and selection appear slower off this branch. Maybe one or a few dependencies haven't been updated properly. Oh wait! Yeah, this branch is still using the `regl-scatter2d@2.1.17`. I push a commit linking it to `regl-scatter2d@master`.

**Comment by etpinard** [comment]
with d972023, the first-render improvements are even better, I'm getting <1000ms first-render time at 1e6 pts which outstanding :tada: 

That said, yeah I'm still experiencing slow drag and selections. @dfcreative maybe we'll need to update parts of update routine.

**Comment by etpinard** [comment]
@dfcreative I made https://github.com/plotly/plotly.js/compare/scattergl-dry-convert...point-cluster-dev which I think captures of the required update in scattergl and scatterpolargl of this branch and applied them to #2492 to avoid conflicts down the road. I suggest closing this PR and working off `point-cluster-dev`.

**Comment by etpinard** [comment]
@dfcreative Here are some problems I noticed off https://github.com/plotly/plotly.js/compare/scattergl-dry-convert...point-cluster-dev :

- gl2d_12, gl2d_14, gl2d_axes_range_type, gl2d_error_bars_log, gl2d_layout_image, are erroring out with:

image

- gl2d_marker_symbols errors out with:

image

- most other mocks produces a diff (see: https://github.com/plotly/plotly.js/commit/badbd59a57fbc9fb2412495926a54ae10ab30486). With the exception of gl2d_scatter-color-clustering all those diff appear to be sub-pixel glitches :ok_hand: . That said, the marker colors in gl2d_scatter-color-clustering are clearly different. @dfcreative was that on-purpose?

**Comment by etpinard** [comment]
as for the jasmine tests, a few tests (6 in total) are failing due to the `TypeError: Cannot read property 'offsets' of undefined` problem mentioned :arrow_heading_up: . No other issues were detected so far.

**Pull request closed without merging**